File: Django_Version/uri_crawler/crawler_provider/ws_problems_data.py
# -*- coding: utf-8 -*-
"""
Created on Fri Nov  3 13:40:06 2017

@author: lucas
"""
from config import *
from utils import *

from urllib2 import urlopen
from bs4 import BeautifulSoup
import re

def get_num_pages(problems_module_address=PROBLEMS_MODULE_ADDRESS):
    html = urlopen(problems_module_address)
    bsObj = BeautifulSoup(html)#, 'lxml'

    num_pages_text = bsObj.find("div", {"id" : "table-info"}).text

    try:
        num_pages = int(re.search('[0-9]+$', num_pages_text).group(0))
    except AttributeError:
        num_pages = 1

    return num_pages

def craw_problems():
    PROBLEMS = []
    print("craw_problems()")
    num_pages = get_num_pages();
    for i in range(1,num_pages+1):
        print("Checking page: " + str(i) + ".. of " + str(num_pages))
        page_idx_address = PROBLEMS_MODULE_ADDRESS + "?page=" + str(i)
        html = urlopen(page_idx_address)
        bsObj = BeautifulSoup(html)#, 'lxml'
        prob_table_data = bsObj.find("table").tbody.findAll("tr")
        
        for prob in prob_table_data:
            identifier = prob.findAll("td")[0].a.text.strip()
            name = prob.findAll("td")[2].a.text
            category_text = prob.findAll("td")[3].a.text
            solved_by_x_users = re.sub('\.', '', prob.findAll("td")[4].text.strip())
            level = prob.findAll("td")[5].text.strip()
            
            categories = ["Iniciante", "Ad-Hoc", "Strings", "Estruturas e Bibliotecas", "Matemática", "Paradigmas", "Grafos", "Geometria Computacional", "SQL"]

            try:
                category = categories.index(category_text) + 1
            except ValueError:
                category = 10

            #print(identifier + "\t" + name + "\t" + category_text + "\t" + solved_by_x_users + "\t" + level +  "\n")
            category = str(category)

            PROBLEMS.append({
                "id":identifier,
                "name": name,
                "category":category,
                "solved":solved_by_x_users,
                "level":level
                })

    return PROBLEMS

File: PHP_Version/services/crawler_provider/ws_problems_data.php
<?php
        
    include_once 'config.php';
    
    function get_num_pages_problems($problems_module_address=PROBLEMS_MODULE_ADDRESS){
    	$doc = new DomDocument;
    	$doc->validateOnParse = true;
    	$doc->loadHtml(file_get_contents($problems_module_address));

    	$num_pages_elem = $doc->getElementById('table-info');
    	if($num_pages_elem != NULL){
    		$matches = array();
    		$num_pages_document = $num_pages_elem->nodeValue;
    		preg_match('/[0-9]+$/', $num_pages_document, $matches);
    		$num_pages = $matches[0];
    	}
    	else
    		$num_pages = 1;

    	return $num_pages;
    }

    function crawl_problems(){
		$PROBLEMS = array();
		//print_r("craw problems() <br>");
		$num_pages = get_num_pages_problems();
		for($i = 1; $i <= $num_pages; $i++){
			//print_r("Checking page: " . $i . ".. of " . $num_pages . "<br>");
			$page_idx_address = (PROBLEMS_MODULE_ADDRESS . "?page=" . $i);

			$doc = new DomDocument;
			$doc->validateOnParse = true;
			$doc->loadHtml(file_get_contents($page_idx_address));

			$table = $doc->getElementsByTagName('table')[0];

			$registers = $table->getElementsByTagName('tr');

			foreach ($registers as $reg) {
				$data = $reg->getElementsByTagName('td');

				$id = str_replace(' ', '', $data[0]->nodeValue); // problem_id
				$name = trim($data[2]->nodeValue); // problem_name
				$category_text = trim($data[3]->nodeValue); // problem_category
				$solved_by_x_users = str_replace(' ', '',$data[4]->nodeValue); // problem_solved_times
				$level = str_replace(' ', '',$data[5]->nodeValue); // problem_level

				//echo "->" . $category_text . "<-";

				$categories = ["Iniciante", "Ad-Hoc", "Strings", "Estruturas e Bibliotecas", "Matemática", "Paradigmas", "Grafos", "Geometria Computacional", "SQL"];

				//$category  = array_search($category_text, $categories) - 1;

				$PROB["id"] = $id;
				$PROB["name"] = $name;
				$PROB["category"] = $category_text;
				$PROB["solved"] = $solved_by_x_users;
				$PROB["level"] = $level;

				$PROBLEMS[$id] = $PROB;

				//array_push($PROBLEMS, $PROB);
			}
		}

		return $PROBLEMS;
    }
    
    set_time_limit(0);
    //var_dump(crawl_problems());

?>